File: packages/gbnf-experiment/data/20260909T035041Z_2d7fbef6/ported_implementation/gbnf/grammar_graph/print.py
from __future__ import annotations

from ..utils.js import from_char_code
from .colorize import Color
from .get_parent_stack_id import get_parent_stack_id
from .type_guards import is_range, is_rule_char, is_rule_ref


def print_graph_pointer(pointer, colorize) -> str:
    return colorize(f"*{get_parent_stack_id(pointer, colorize)}", Color.RED)


def print_graph_node(node, colorize, pointers=None, show_position: bool = False) -> str:
    col = colorize
    rule = node.rule

    parts: list[str] = []
    if show_position:
        parts.extend(
            [
                col("{", Color.BLUE),
                col(node.id, Color.GRAY),
                col("}", Color.BLUE),
            ]
        )
    if is_rule_char(rule):
        parts.extend(
            [
                col("[", Color.GRAY),
                col(
                    "".join(
                        # a range renders as an array in JS, which `join` stringifies
                        # with commas
                        ",".join(col(from_char_code(val), Color.YELLOW) for val in v)
                        if is_range(v)
                        else _get_char(v)
                        for v in rule.value
                    ),
                    Color.YELLOW,
                ),
                col("]", Color.GRAY),
            ]
        )
    elif is_rule_ref(rule):
        parts.append(
            col("Ref(", Color.GRAY) + col(f"{rule.value}", Color.GREEN) + col(")", Color.GRAY)
        )
    else:
        parts.append(col(rule.type.value, Color.YELLOW))

    if pointers:
        for pointer in pointers:
            pointer_parts: list[str] = []
            if pointer.node is node:
                pointer_parts.append(pointer.print(colorize=col))
            if pointer_parts:
                parts.append(col("[", Color.GRAY))
                parts.extend(pointer_parts)
                parts.append(col("]", Color.GRAY))

    rest = (
        node.next.print(pointers=pointers, colorize=col, show_position=show_position)
        if node.next is not None
        else None
    )
    return col("-> ", Color.GRAY).join(part for part in ["".join(parts), rest] if part)


def _get_char(char_code: int) -> str:
    char = from_char_code(char_code)
    if char == "\n":
        return "\\n"
    return char

File: proposed-projects/round-trip-experiment/reverse/20260909T035041Z_2d7fbef6/20260910T064524Z_d6fcf00e/ported_implementation/src/grammar-graph/print.ts
import { Color, Colorize } from './colorize.js';
import { getParentStackId } from './get-parent-stack-id.js';
import type { GenericSet } from './generic-set.js';
import type { GraphNode } from './graph-node.js';
import type { GraphPointer } from './graph-pointer.js';
import { isRange, isRuleChar, isRuleRef } from './type-guards.js';

export const printGraphPointer = (pointer: GraphPointer, colorize: Colorize): string =>
  colorize(`*${getParentStackId(pointer, colorize)}`, Color.RED);

export const printGraphNode = (
  node: GraphNode,
  colorize: Colorize,
  pointers?: GenericSet<GraphPointer, string>,
  showPosition = false,
): string => {
  const col = colorize;
  const { rule } = node;

  const parts: string[] = [];
  if (showPosition) {
    parts.push(col('{', Color.BLUE), col(node.id, Color.GRAY), col('}', Color.BLUE));
  }
  if (isRuleChar(rule)) {
    parts.push(
      col('[', Color.GRAY),
      col(
        rule.value
          .map((v) =>
            // a range renders as an array in JS, which `join` stringifies with commas
            isRange(v)
              ? v.map((val) => col(String.fromCharCode(val), Color.YELLOW)).join(',')
              : getChar(v),
          )
          .join(''),
        Color.YELLOW,
      ),
      col(']', Color.GRAY),
    );
  } else if (isRuleRef(rule)) {
    parts.push(
      col('Ref(', Color.GRAY) + col(`${rule.value}`, Color.GREEN) + col(')', Color.GRAY),
    );
  } else {
    parts.push(col(rule.type, Color.YELLOW));
  }

  if (pointers) {
    for (const pointer of pointers) {
      const pointerParts: string[] = [];
      if (pointer.node === node) {
        pointerParts.push(pointer.print(col));
      }
      if (pointerParts.length) {
        parts.push(col('[', Color.GRAY), ...pointerParts, col(']', Color.GRAY));
      }
    }
  }

  const rest = node.next ? node.next.print(col, pointers, showPosition) : undefined;
  return [parts.join(''), rest].filter(Boolean).join(col('-> ', Color.GRAY));
};

const getChar = (charCode: number): string => {
  const char = String.fromCharCode(charCode);
  if (char === '\n') {
    return '\\n';
  }
  return char;
};
